### PR TITLE
Switch YAML tests to using templated ReadAttribute.

### PR DESCRIPTION
**Pull request by bzbarsky-apple**
This will let us handle reading nullables (and structs once we can
have struct-typed attributes) propely.

#### Problem
YAML tests can't read interesting types (except for lists, which we have some special-case hacks for).

#### Change overview
Switch to using templated ReadAttribute to allow us to do that.

Also fixes the ReadAttribute templates to instantiate most of the work on a per-type basis, not per-attribute basis.

#### Testing
No behavior changes so far, but will be building on top of this to expand testing.

### CODE REVIEW

**Review by woody-apple — APPROVED** [review]

**Review by woody-apple — APPROVED** [review]

File: src/app/zap-templates/templates/app/cluster-objects.zapt [hunk 1]
```diff
@@ -139,9 +139,11 @@ namespace {{asUpperCamelCase label}} {
     {{#if entryType}}
         using Type = {{zapTypeToEncodableClusterObjectType entryType forceNotOptional=true}};
         using DecodableType = {{zapTypeToDecodableClusterObjectType entryType forceNotOptional=true}};
+        using DecodableArgType = {{zapTypeToDecodableClusterObjectType entryType forceNotOptional=true isArgument=true}};
     {{else}}
         using Type = {{zapTypeToEncodableClusterObjectType type forceNotOptional=true}};
         using DecodableType = {{zapTypeToDecodableClusterObjectType type forceNotOptional=true}};
+        using DecodableArgType = {{zapTypeToDecodableClusterObjectType type forceNotOptional=true isArgument=true}};
```
Comment (mrjerryjohns) [inline src/app/zap-templates/templates/app/cluster-objects.zapt:146]: Why is this needed?

Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/cluster-objects.zapt:146]: Because the actual callbacks that are used for Read right now for the ChipClusters consumers take things like `uint8_t` as arguments, not `const uint8_t &`.  So either I need to make the CHIPClusters Read API expect callbacks with those types, which means I need a way of deriving them from AttributeInfo, or I need to change all those callbacks' signatures.

And in choosing between those options, needing to implement a function that takes `const bool &` to call something that will eventually return a bool is pretty odd.

Comment (mrjerryjohns) [reply src/app/zap-templates/templates/app/cluster-objects.zapt:146]: Why not keep it simple and have all types be returned by reference, regardless of type?

That would then permit you to always pass in `const DecodableType &` to CHIPClusters consumers, without requiring us to generate two different `DecodableType` versions.

Put it differently, if chip-tool was an external consumer, how would we have solved it?

Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/cluster-objects.zapt:146]: > Why not keep it simple and have all types be returned by reference, regardless of type? 

We could; it just involves changing a lot more code to make it work (both here and to make other consumers adopt this API), and leads to a less natural-looking API.

Naming `DecodableArgType` something like `DecodableTypeForArgs` might make it clearer, I agree.

**Review by mrjerryjohns — APPROVED** [review]

File: src/app/zap-templates/templates/app/cluster-objects.zapt (hunk 1 above)
Comment (bzbarsky-apple) [reply src/app/zap-templates/templates/app/cluster-objects.zapt:146]: Per Slack discussion: do this for now, then once we get Darwin/Java/etc on the new APIs re-evaluate.

**Comment by bzbarsky-apple** [comment]
@saurabhst @msandstedt @jepenven-silabs @jmartinez-silabs @Damian-Nordic @LuDuda Please take a look?

**Review by saurabhst — APPROVED** [review]
